File: java/test/src/main/java/test/Ice/operations/BatchOnewaysAMI.java
// Copyright (c) ZeroC, Inc.

package test.Ice.operations;

import com.zeroc.Ice.Communicator;
import com.zeroc.Ice.Identity;
import com.zeroc.Ice.ObjectPrx;
import com.zeroc.Ice.Properties;
import com.zeroc.Ice.Util;

import test.Ice.operations.Test.MyClassPrx;

import java.io.PrintWriter;

class BatchOnewaysAMI {
    private static class Callback {
        Callback() {
            _called = false;
        }

        public synchronized void check() {
            while (!_called) {
                try {
                    wait();
                } catch (InterruptedException ex) {}
            }

            _called = false;
        }

        public synchronized void called() {
            assert (!_called);
            _called = true;
            notify();
        }

        private boolean _called;
    }

    private static void test(boolean b) {
        if (!b) {
            throw new RuntimeException();
        }
    }

    static void batchOneways(MyClassPrx p, PrintWriter out) {
        final Communicator communicator = p.ice_getCommunicator();
        final Properties properties = communicator.getProperties();
        final byte[] bs1 = new byte[10 * 1024];

        MyClassPrx batch = p.ice_batchOneway();
        batch.ice_flushBatchRequestsAsync().join(); // Empty flush

        {
            test(batch.ice_flushBatchRequestsAsync().isDone()); // Empty flush
            test(
                Util.getInvocationFuture(batch.ice_flushBatchRequestsAsync())
                    .isSent()); // Empty flush
            test(
                Util.getInvocationFuture(batch.ice_flushBatchRequestsAsync())
                    .sentSynchronously()); // Empty flush
        }

        for (int i = 0; i < 30; i++) {
            batch.opByteSOnewayAsync(bs1)
                .whenComplete(
                    (result, ex) -> {
                        test(ex == null);
                    });
        }

        int count = 0;
        while (count < 27 /* 3*9 requests auto-flushed */) {
            count += p.opByteSOnewayCallCount();
            try {
                Thread.sleep(10);
            } catch (InterruptedException ex) {}
        }

        final boolean bluetooth =
            properties.getIceProperty("Ice.Default.Protocol").indexOf("bt") == 0;
        if (batch.ice_getConnection() != null && !bluetooth) {
            MyClassPrx batch2 = p.ice_batchOneway();

            batch.ice_pingAsync();
            batch2.ice_pingAsync();
            batch.ice_flushBatchRequestsAsync().join();
            batch.ice_getConnection().close();
            batch.ice_pingAsync();
            batch2.ice_pingAsync();

            batch.ice_getConnection();
            batch2.ice_getConnection();

            batch.ice_pingAsync();
            batch.ice_getConnection().close();
            test(!batch.ice_pingAsync().isCompletedExceptionally());
            test(!batch2.ice_pingAsync().isCompletedExceptionally());
        }

        Identity identity = new Identity();
        identity.name = "invalid";
        ObjectPrx batch3 = batch.ice_identity(identity);
        batch3.ice_pingAsync();
        batch3.ice_flushBatchRequestsAsync().join();

        // Make sure that a bogus batch request doesn't cause troubles to other ones.
        batch3.ice_pingAsync();
        batch.ice_pingAsync();
        batch.ice_flushBatchRequestsAsync().join();
        batch.ice_pingAsync();
    }

    private BatchOnewaysAMI() {}
}

File: matlab/test/Ice/operations/BatchOnewaysAMI.m
%{
**********************************************************************

Copyright (c) 2003-2017 ZeroC, Inc. All rights reserved.

This copy of Ice is licensed to you under the terms described in the
ICE_LICENSE file included in this distribution.

**********************************************************************
%}

classdef BatchOnewaysAMI
    methods(Static)
        function batchOneways(app, p)
            communicator = app.communicator();
            properties = communicator.getProperties();
            bs1 = zeros(1, 10  * 1024);

            batch = p.ice_batchOneway();
            future = batch.ice_flushBatchRequestsAsync(); % Empty flush
            future.wait();
            assert(strcmp(future.State, 'finished'));

            futures = {};
            for i = 1:30
                futures{i} = batch.opByteSOnewayAsync(bs1);
            end
            for i = 1:30
                assert(futures{i}.wait());
            end

            count = 0;
            while count < 27 % 3 * 9 requests auto-flushed.
                count = count + p.opByteSOnewayCallCount();
                pause(0.1);
            end

            if ~isempty(batch.ice_getConnection())
                batch2 = p.ice_batchOneway();

                batch.ice_pingAsync();
                batch2.ice_pingAsync();
                assert(batch.ice_flushBatchRequestsAsync().wait());
                batch.ice_getConnection().close(Ice.ConnectionClose.GracefullyWithWait);
                batch.ice_pingAsync();
                batch2.ice_pingAsync();

                batch.ice_getConnection();
                batch2.ice_getConnection();

                batch.ice_pingAsync();
                batch.ice_getConnection().close(Ice.ConnectionClose.GracefullyWithWait);
                assert(batch.ice_pingAsync().wait());
                assert(batch2.ice_pingAsync().wait());
            end

            identity = Ice.Identity();
            identity.name = 'invalid';
            batch3 = batch.ice_identity(identity);
            batch3.ice_pingAsync();
            assert(batch3.ice_flushBatchRequestsAsync().wait());

            % Make sure that a bogus batch request doesn't cause troubles to other ones.
            batch3.ice_pingAsync();
            batch.ice_pingAsync();
            assert(batch.ice_flushBatchRequestsAsync().wait());
            batch.ice_pingAsync();
        end
    end
end
